### PR TITLE
Initial CLI Version of `cfssl scan`

### PR DESCRIPTION
**Pull request by jacobhaven**
Adds new `scan` command, that performs a variety of connectivity and TLS protocol tests.

Unfortunately, some of these require importing the go standard library's `crypto/tls` package's source in order to access private members. These are used in `scan/tls/cfsslscan_*.go` files to export additional functionality not supplied by the `crypto/tls` API.

Support for the standard CFSSL HTTP API is forthcoming, but comments on this initial structure might be useful first.

### CODE REVIEW

File: scan/scan_common.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+// Scanner describes a type of scan to perform on a host.
+type Scanner struct {
+	Name        string
+	Description string
+	scan        func(host string) (Grade, Output, error)
+}
+
+// Scan defines the scan to be performed on the given host.
+func (s *Scanner) Scan(host string) (grade Grade, output Output, err error) {
```
Comment (lziest) [inline scan/scan_common.go]: Looks good. One style issue: For exported function, we keep the return list nameless, so the doc will be consistent with standard GoDoc. Also, it makes each return statement declare exactly what should be returned. It's much easier to maintain afterwards, and reduces potential bugs.

File: scan/tls_ciphersuite_scanner.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package scan
+
+import (
+	"errors"
+	"net"
+
+	"github.com/jacobhaven/cipherscan/tls"
```
Comment (lziest) [inline scan/tls_ciphersuite_scanner.go]: Some compilation error regarding this package.

File: scan/tls_session_scanner.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	addrs, err := net.LookupHost(host)
+	if err != nil {
+		return
+	}
+	sessions := newSessionCache()
+	output = sessions
+	config := &tls.Config{ClientSessionCache: sessions}
+	for _, addr := range addrs {
+		var conn *tls.Conn
+		conn, err = tls.Dial(Network, addr, config)
```
Comment (lziest) [inline scan/tls_session_scanner.go]: I recommend to use DialWithDialer, so we can control the timeout and easy to tweak. It's otherwise the system default timeout, which could be like 30 seconds.

File: cfssl_scan.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	"regexp"
+
+	"github.com/cloudflare/cfssl/scan"
+)
+
+var scanUsageText = `cfssl scan -- scan a host for issues
+
+Usage of scan:
+        cfssl scan [-short] [SCANFAMILY] HOST
+        cfssl scan [-short] [-scanner regexp] HOST
```
Comment (grittygrease) [inline cfssl_scan.go]: Please put an example here. The documentation is confusing, is SCANFAMILY a regexp? What is regexp matching against? What does [SCANFAMILY]* represent?

File: scan/tls_ciphersuite_scanner.go (hunk 2 above)
Comment (grittygrease) [inline scan/tls_ciphersuite_scanner.go]: Don't we want the version from cfssl/scan/tls here?

Comment (jacobhaven) [inline scan/tls_ciphersuite_scanner.go]: Ah, well that would explain something. \*updating\*

File: cfssl_scan.go (hunk 4 above)
Comment (jacobhaven) [inline cfssl_scan.go]: I've refactored this to be a normal argument `-scanfamily regexp`

File: cfssl.go [hunk 5]
```diff
@@ -67,6 +67,9 @@ var Config struct {
 	domain            string
 	ip                string
 	remote            string
+	short             bool
+	list              bool
+	scanner           string
```
Comment (grittygrease) [inline cfssl.go]: Need to register flags and command (CLIScan) below.

File: scan/connectivity_scanner.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	Description: "Check for basic connectivity with the host",
+	Scanners: []*Scanner{
+		DNSLookupScanner,
+		TCPDialScanner,
+		TLSDialScanner,
+	},
+}
+
+// DNSLookupScanner tests that DNS resolution of the host returns at least one address
+var DNSLookupScanner = &Scanner{
```
Comment (kisom) [inline scan/connectivity_scanner.go]: Should use the field names in initialisation.

File: scan/connectivity_scanner.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package scan
+
+import "net"
+
+// ConnectivityScans contains scanners testing basic connectivity to the host
+var ConnectivityScanners = &Family{
+	Name:        "ConnectivityScanners",
```
Comment (kisom) [inline scan/connectivity_scanner.go]: Style note: if Family denotes a family of scanners, you can probably drop the "Scanners" part of the name. if this is the name that the user selects the family with on the commandline, I would just use "connectivity" to make it easier and consistent with the other args.

File: scan/connectivity_scanner.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func dnsLookupScan(host string) (grade Grade, output Output, err error) {
+	var conn net.Conn
+	addrs, err := net.LookupHost(host)
+	if err != nil || len(addrs) == 0 {
+		output = []string{}
+		return
+	}
+	defer conn.Close()
```
Comment (kisom) [inline scan/connectivity_scanner.go]: This shouldn't be deferred; the conn is no longer in use and should be closed right away.

File: scan/scan_common.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	// Legacy describes a host with non-ideal configuration that maintains support for legacy clients.
+	Legacy
+	// Good describes host performing the expected state-of-the-art.
+	Good
+)
+
+// Output is the result of a Scan, to be stored for potential use by later Scanners.
+type Output interface{}
+
+// Scanner describes a type of scan to perform on a host.
```
Comment (kisom) [inline scan/scan_common.go]: One thing I'm trying to move to in CFSSL is to make sure structure fields are documented as well.

File: scan/scan_common.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	output = outputs[len(outputs)-1]
+	return
+}
+func (h *scanHistory) Store(scannerName, host string, output Output) {
+	key := scannerHostPair{scannerName, host}
+	h.history[key] = append(h.history[key], output)
+}
+
+// Family defines a set of related scans meant to be run together in sequence
+type Family struct {
```
Comment (kisom) [inline scan/scan_common.go]: These field comments are good. As a Go style note though, they should be complete sentences.

File: scan/connectivity_scanner.go (hunk 8 above)
Comment (jacobhaven) [inline scan/connectivity_scanner.go]: I didn't , this was a a relic of before this test was split from another one (It actually caused a bug and was removed)

File: cfssl_scan.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        cfssl scan [-list] [SCANFAMILY]*
+
+Arguments:
+        HOST:    Host to scan (including port)
+        SCANFAMILY:	 Regexp selecting family of related scanners to run
```
Comment (grittygrease) [inline cfssl_scan.go]: Inconsistency with spaces v. tabs.

File: cfssl_scan.go [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+						fmt.Printf("\t%s:\n\t\t%s", scanner.Name, scanner.Description)
+					}
+				}
+			}
+		}
+	} else {
+		if Config.scanner != "" {
+			scannerRegexp = regexp.MustCompile(Config.scanner)
+		}
+		for len(args) > 1 {
```
Comment (grittygrease) [inline cfssl_scan.go]: This logic needs some work. calling cfssl scan with no arguments causes scanFamily.Name to be nil and 
			matches := scanFamilyRegexp.MatchString(scanFamily.Name)
to cause a panic.

File: scan/tls/alert.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2009 The Go Authors. All rights reserved.
```
Comment (grittygrease) [inline scan/tls/alert.go]: It would be very helpful to have some documentation here. At the minimum we need a revision of the original code this was copied from and a diff to see which changes are custom.

Comment (grittygrease) [inline scan/tls/alert.go]: It might even make sense to leave this as its own repository (forked from crypto/tls) so that we can track the diffs and keep it up to date with Go's tls changes.

**Comment by kisom** [comment]
You've checked in the CFSSL binary. Please remove this (and make sure the commits are squashed so that it doesn't appear in any commits).

File: scan/tls_handshake_scanners.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	tcpConn, err := net.Dial(Network, host)
+	if err != nil {
+		return
+	}
+	conn := tls.Client(tcpConn, &tls.Config{ServerName: host, CipherSuites: ciphers})
+	defer conn.Close()
+	serverCipher, serverVersion, err = conn.SayHello()
+	for _, cipherID := range ciphers {
+		if serverCipher == cipherID {
+			err = errors.New("Server negotiated ciphersuite we didn't send.")
```
Comment (kisom) [inline scan/tls_handshake_scanners.go]: As a style note, Go error strings are generally not capitalised and don't end in punctuation.

File: scan/tls/alert.go (hunk 13 above)
Comment (jacobhaven) [inline scan/tls/alert.go]: This isn't a bad idea. To set it up as a submodule. I don't think we'll need to make any changes to the original `crypto/tls` files. I've just needed to access some functionality in creating the helper functions in `tls/cfsslscan_*`

Comment (grittygrease) [inline scan/tls/alert.go]: No need to use a submodule in Go, you can just include from Github. For safety, the module name could be changed to scan-tls, cf-tls or something like that.

File: cfssl_scan.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+					if err != nil {
+						return
+					}
+				}
+			}
+		}
+	}
+	return
+}
+func scanMain(args []string) (err error) {
```
Comment (jgrahamc) [inline cfssl_scan.go]: You've mixed the user of a named return argument and returns with arguments. Choose one or the other. I'd suggest making the returned value just 'error' and using explicit 'return foo'.

File: cfssl_scan.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        cfssl scan -list [-short] [-scanfamily regexp] [-scanner regexp]
+
+Arguments:
+        HOST:    Host(s) to scan (including port)
+Flags:
+`
+var scanFlags = []string{"short", "list", "family", "scanner"}
+var scanFamilyRegexp, scannerRegexp *regexp.Regexp
+
+func regexpLoop(familyFunc func(*scan.Family) error, scannerFunc func(*scan.Scanner) error) (err error) {
```
Comment (jgrahamc) [inline cfssl_scan.go]: I think its worth documenting these arguments as they are rather complex and it's not immediately clear what's expected of a familyFunc or a scannerFunc.

File: scan/connectivity.go [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// dnsLookupScan tests that DNS resolution of the host returns at least one address
+func dnsLookupScan(host string) (grade Grade, output Output, err error) {
+	host, _, err = net.SplitHostPort(host)
+	if err != nil {
+		return
+	}
+	addrs, err := net.LookupHost(host)
+	if err != nil || len(addrs) == 0 {
```
Comment (jgrahamc) [inline scan/connectivity.go]: If err == nil but len(addrs) == 0 this will return nil. I think you want something else.

File: cfssl.go [hunk 18]
```diff
@@ -94,6 +98,10 @@ func registerFlags() {
 	cfsslFlagSet.StringVar(&Config.domain, "domain", "", "remote server domain name")
 	cfsslFlagSet.StringVar(&Config.ip, "ip", "", "remote server ip")
 	cfsslFlagSet.StringVar(&Config.remote, "remote", "", "remote CFSSL server")
+	cfsslFlagSet.BoolVar(&Config.short, "short", false, "so not run long/intensive scans")
```
Comment (jgrahamc) [inline cfssl.go]: Not sure what 'so no run long/intensive scans' means.

File: cfssl.go [hunk 19]
```diff
@@ -67,6 +67,10 @@ var Config struct {
 	domain            string
 	ip                string
 	remote            string
+	short             bool
+	list              bool
+	family            string
```
Comment (jgrahamc) [inline cfssl.go]: These need documenting. It's very unclear how family and scanner interact.

File: scan/scan_common.go [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+func (s *Scanner) String() string {
+	ret := fmt.Sprintf("\t%s", s.Name)
+	if !Short {
+		ret += fmt.Sprintf(":\n\t\t%s", s.Description)
+	}
+	return ret
+}
+
+type scannerHostPair struct{ scannerName, host string }
+type scanHistory struct {
```
Comment (jgrahamc) [inline scan/scan_common.go]: If all you have in scanHistory is this map then just do
```
type scanHistory map[scannerHostPair][]Output
```
and simplify your life.

File: scan/tls_handshake.go [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+func (ciphers cipherList) String() (list string) {
+	for _, cipher := range ciphers {
+		list += fmt.Sprintf("%s(%#04x), ", tls.CipherSuites[cipher], cipher)
+	}
+	if len(list) > 2 {
+		list = list[:len(list)-2]
+	}
+	return list
+}
+func (ciphers cipherList) Append(cipher uint16) {
```
Comment (jgrahamc) [inline scan/tls_handshake.go]: This function seems a bit pointless.

File: scan/tls_session.go [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+		{
+			"SessionResumeScanner",
+			"Host is able to resume sessions accross all addresses.",
+			sessionResumeScan,
+		},
+	},
+}
+
+// sessionCache is a mock ClientSessionCache used to record all stores (i.e. new session tickets)
+type sessionCache struct {
```
Comment (jgrahamc) [inline scan/tls_session.go]: This could just be 
```
type sessionCache map[string][]*tls.ClientSessionState
```

File: scan/tls_session.go [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+	cache map[string][]*tls.ClientSessionState
+}
+
+func newSessionCache() *sessionCache {
```
Comment (jgrahamc) [inline scan/tls_session.go]: If you make the change above then this function is unnecessary and you can just do 
```
make(sessionCache)
```
whenever one is needed.

File: scan/tls_session.go [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+		return
+	}
+	mostRecent = states[len(states)-1]
+	return
+}
+func (s *sessionCache) Put(sessionKey string, cs *tls.ClientSessionState) {
+	s.cache[sessionKey] = append(s.cache[sessionKey], cs)
+}
+
+// SessionResumeScan tests that host is able to resume sessions accross all addresses.
```
Comment (jgrahamc) [inline scan/tls_session.go]: across

File: scan/tls_session.go [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return
+	}
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return
+	}
+	sessions := newSessionCache()
+	output = sessions
+	config := &tls.Config{ClientSessionCache: sessions, InsecureSkipVerify: true}
```
Comment (jgrahamc) [inline scan/tls_session.go]: Why 
```
InsecureSkipVerify: true
```
? I think that needs documenting.

File: scan/tls_session.go [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+	for _, ip := range ips {
+		var addr string
+		if ip.To4() != nil {
+			addr = ip.String()
+		} else {
+			addr = "[" + ip.String() + "]"
+		}
+		if len(port) == 0 {
```
Comment (jgrahamc) [inline scan/tls_session.go]: Why not use
```
port == ""
```
instead of
```
len(port) == 0
```

File: scan/tls_session.go [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+			port = ":443"
+		} else if port[0] != ':' {
+			port = ":" + port
+		}
+		fmt.Println()
+		var conn *tls.Conn
+		conn, err = tls.Dial(Network, addr+port, config)
```
Comment (jgrahamc) [inline scan/tls_session.go]: There's actual a function called
```
net.JoinHostPort
```
http://golang.org/pkg/net/#JoinHostPort

File: scan/tls_session.go [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+		if err != nil {
+			return
+		}
+		conn.Close()
+		if len(sessions.cache) != 1 {
+			return
```
Comment (jgrahamc) [inline scan/tls_session.go]: Is it OK that the error returned here is nil?

File: cfssl_scan.go (hunk 16 above)
Comment (jacobhaven) [inline cfssl_scan.go]: Updated with:
```
// regexpLoop iterates through each scan Family and Scanner registered in scan.AllFamilies.
// familyFunc is run on each Family with a name matching the family flag's regexp,
// then scannerFunc is run on each Scanner in that Family that matches the scanner flag's regexp
```

File: cfssl.go (hunk 19 above)
Comment (jacobhaven) [inline cfssl.go]: Essentially you should probably only use one or the other. Using both takes the intersection. How does this sound?
```
  -family="": regular expression selecting specific scan families to run (defaults to all)
  -scanner="": regular expression selecting specific scanners to run (defaults to all)
```

File: cfssl.go (hunk 18 above)
Comment (jacobhaven) [inline cfssl.go]: It was intended as "don't run long and/or intensive scans". But, preferably we won't have those, or you can select to not run the entire suite. The `short` argument has been removed, and and replaced with a `verbose` output argument.

File: cfssl_scan.go (hunk 15 above)
Comment (jacobhaven) [inline cfssl_scan.go]: Most of the other cfssl Mains use `(err error)`, so I'll stick with that, but fix the `return err`/`return nil` instances

File: scan/connectivity.go (hunk 17 above)
Comment (jacobhaven) [inline scan/connectivity.go]: Good point. Split into two separate errors.

File: scan/scan_common.go (hunk 1 above)
Comment (jacobhaven) [inline scan/scan_common.go]: Good point, done.

File: scan/scan_common.go (hunk 9 above)
Comment (jacobhaven) [inline scan/scan_common.go]: Done.

File: scan/scan_common.go (hunk 20 above)
Comment (jacobhaven) [inline scan/scan_common.go]: True. At first I thought there might be more included in there. I've refactored a large part of this.

File: scan/tls_handshake.go (hunk 21 above)
Comment (jacobhaven) [inline scan/tls_handshake.go]: Fair :)

File: scan/tls_session.go (hunk 22 above)
Comment (jacobhaven) [inline scan/tls_session.go]: I've actually removed this type entirely and used `tls.NewLRUClientSessionCache()` instead

File: scan/tls_session.go (hunk 25 above)
Comment (jacobhaven) [inline scan/tls_session.go]: I suppose there's no reason not to verify here, so I will.

File: scan/tls_session.go (hunk 26 above)
Comment (jacobhaven) [inline scan/tls_session.go]: Actually, port should *always* be populated (already defaulted to 443 if empty), so this is completely unnecessary.

**Comment by jacobhaven** [comment]
Rebased on new `cmd` and `cli` structure and cleaned up commit log, added new scan families for secure HTTP headers and (extending the scope of the tool), and scanner to scrape CIDR IP ranges for new Intermediate CAs to update [cfssl_trust](https://github.com/cloudflare/cfssl_trust) for Bundler.

**Comment by grittygrease** [comment]
Unfortunately, some of the go code copied from the standard library does not follow the official style guidelines and fails go vet. Please check the travis output and fix.

File: scan/https.go [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+		fing := base64.StdEncoding.EncodeToString(digest[:])
+		if strings.Contains(hpkpHeader, fing) {
+			output = fingerprint(fing)
+			grade = Good
+		}
+	}
+	return
+}
+
+// hstsScan tests that the host serves correct HPKP headers.
```
Comment (grittygrease) [inline scan/https.go]: Do you also want to check for expiry time here? Short period HSTS is less valuable than long period HSTS.

Comment (jacobhaven) [inline scan/https.go]: Possibly. I mainly couldn't find an easy way to parse the header values.

**Comment by grittygrease** [comment]
@jacobhaven Can you add tests for these packages? Even a nearly empty test file in each package makes sure that the overall coverage takes this code into account.

File: scan/scan_common.go [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+const (
+	// Bad describes a host with serious misconfiguration or vulnerability.
+	Bad Grade = iota
+	// Legacy describes a host with non-ideal configuration that maintains support for legacy clients.
+	Legacy
+	// Good describes host performing the expected state-of-the-art.
+	Good
+)
+
+func (g Grade) String() string {
```
Comment (lziest) [inline scan/scan_common.go]: Minor, add doc.

File: scan/scan_common.go [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+func (s *Scanner) Scan(host string) (Grade, Output, error) {
+	grade, output, err := s.scan(host)
+	if err != nil {
+		log.Infof("%s: %s", s.Name, err)
+		return grade, output, err
+	}
+	History.store(s.Name, host, output)
+	return grade, output, err
+}
+func (s *Scanner) String() string {
```
Comment (lziest) [inline scan/scan_common.go]: Minor, add doc and a blank line

File: scan/scan_common.go [hunk 32]
```diff
@@ -0,0 +1,142 @@
...
+type Family struct {
+	// Name is a short name for the Family.
+	Name string
+	// Description gives a short description of the scans performed on the host.
+	Description string
+	// Scanners is a list of scanners that are to be run in sequence.
+	Scanners []*Scanner
+}
+
+func (f *Family) String() string {
```
Comment (lziest) [inline scan/scan_common.go]: Minor, add doc.

File: scan/tls_handshake.go [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+	var cipherID uint16
+	for cipherIndex, cipherID = range ciphers {
+		if serverCipher == cipherID {
+			return
+		}
+	}
+	err = fmt.Errorf("server negotiated ciphersuite we didn't send: %s", tls.CipherSuites[serverCipher])
+	return
+}
+func allCiphersIDs() []uint16 {
```
Comment (lziest) [inline scan/tls_handshake.go]: Add blank lines for style consistency.

File: cli/scan/scan.go [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+func indentPrintln(block string, indentLevel int) {
+	w := tabwriter.NewWriter(os.Stdout, 0, 4, 4, ' ', 0)
+	scanner := bufio.NewScanner(strings.NewReader(block))
+	for scanner.Scan() {
+		fmt.Fprintf(w, "%s%s\n", strings.Repeat("\t", indentLevel), scanner.Text())
+	}
+	w.Flush()
+}
+func scanMain(args []string, c cli.Config) (err error) {
+	familyRegexp := regexp.MustCompile(c.Family)
```
Comment (lziest) [inline cli/scan/scan.go]: Minor issue: do we want to panic here if the input is not valid regex? Just use Compile and throw error here?

File: cli/cli.go [hunk 35]
```diff
@@ -94,6 +97,9 @@ func registerFlags(c *Config, f *flag.FlagSet) {
 	f.StringVar(&c.IP, "ip", "", "remote server ip")
 	f.StringVar(&c.Remote, "remote", "", "remote CFSSL server")
 	f.StringVar(&c.Label, "label", "", "key label to use in remote CFSSL server")
+	f.BoolVar(&c.List, "list", false, "list possible scanners")
+	f.StringVar(&c.Family, "family", "", "family regular expression")
```
Comment (lziest) [inline cli/cli.go]: I prefer 'scanner family regular expression'

**Comment by lziest** [comment]
Other than those minor issues, LGTM

**Comment by grittygrease** [comment]
Once Zi's comments are addressed and the _test.go files are added, I think this is ready to merge.

File: cli/scan/scan.go (hunk 34 above)
Comment (kisom) [inline cli/scan/scan.go]: I think that we should use Compile,m and return an error instead of panicking.

File: scan/connectivity.go [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+func (addrs lookupAddrs) Describe() string {
+	return "List of host's addresses returned by DNS lookup"
+}
+
+// dnsLookupScan tests that DNS resolution of the host returns at least one address
+func dnsLookupScan(host string) (grade Grade, output Output, err error) {
+	host, _, err = net.SplitHostPort(host)
+	if err != nil {
+		return
+	}
```
Comment (kisom) [inline scan/connectivity.go]: It's a style nit, but I find the code is more readable if you put a blank line here (and L53).

File: scan/https.go [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+			}
+			return conn, err
+		},
+	}
+	client := http.Client{Transport: tr}
+	resp, err := client.Get("https://" + host)
+	if err != nil {
+		return
+	}
+	hpkpHeader := resp.Header.Get("Public-Key-Pins")
```
Comment (kisom) [inline scan/https.go]: Style nit: I think it's more readable if you put a newline between the '}' and this line.

File: scan/tls_handshake.go [hunk 38]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		return
+	}
+	conn := tls.Client(tcpConn, &tls.Config{MinVersion: vers, MaxVersion: vers, ServerName: host, CipherSuites: ciphers, SessionTicketsDisabled: true, ClientSessionCache: nil})
+	serverCipher, serverVersion, err := conn.SayHello()
+	conn.Close()
+	if err != nil {
+		return
+	}
+	if serverVersion != vers {
```
Comment (kisom) [inline scan/tls_handshake.go]: I'd add a blank line here.

File: scan/tls_handshake.go [hunk 39]
```diff
@@ -0,0 +1,111 @@
...
+		err = fmt.Errorf("server negotiated protocol version we didn't send: %s", tls.Versions[serverVersion])
+		return
+	}
+	var cipherID uint16
```
Comment (kisom) [inline scan/tls_handshake.go]: and probably here, too.

File: scan/tls_handshake.go [hunk 40]
```diff
@@ -0,0 +1,111 @@
...
+		ciphers = append(ciphers, cipherID)
+	}
+	return ciphers
+}
+
+type cipherVersions struct {
+	cipherID uint16
+	versions []uint16
+}
+type cipherVersionList []cipherVersions
```
Comment (kisom) [inline scan/tls_handshake.go]: Should add a blank line.

File: scan/tls_handshake.go [hunk 41]
```diff
@@ -0,0 +1,111 @@
...
+	for i, c := range cvList {
+		versStrings := make([]string, len(c.versions))
+		for j, vers := range c.versions {
+			versStrings[j] = tls.Versions[vers]
+		}
+		cvStrings[i] = fmt.Sprintf("%s\t%s", tls.CipherSuites[c.cipherID], strings.Join(versStrings, ", "))
+	}
+	return strings.Join(cvStrings, "\n")
+}
+func (cvList cipherVersionList) Describe() string {
```
Comment (kisom) [inline scan/tls_handshake.go]: Blank line above this.

File: scan/tls_session.go [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+		return
+	}
+	config := &tls.Config{ClientSessionCache: tls.NewLRUClientSessionCache(1), ServerName: hostname}
+	var conn *tls.Conn
+	conn, err = tls.Dial(Network, host, config)
+	if err != nil {
+		return
+	}
+	conn.Close()
+	for _, ip := range ips {
```
Comment (kisom) [inline scan/tls_session.go]: I'd add a blank line about this one (after the `conn.Close()`).

**Comment by kisom** [comment]
Only nits I had were stylistic; other than those, LGTM.

**Comment by kisom** [comment]
Also, @jacobhaven you'll need to resolve the merge conflicts before I can merge this.

**Comment by grittygrease** [comment]
@jacobhaven Can you address Kyle's comments so we can merge and move on to adding an API?

File: cli/config.go [hunk 43]
```diff
@@ -72,6 +75,9 @@ func registerFlags(c *Config, f *flag.FlagSet) {
 	f.IntVar(&c.Reason, "reason", 0, "Reason code for revocation")
 	f.StringVar(&c.RevokedAt, "revoked-at", "now", "Date of revocation (YYYY-MM-DD)")
 	f.Int64Var(&c.Interval, "interval", int64(4*helpers.OneDay), "Interval between OCSP updates, in seconds (default: 4 days)")
+	f.BoolVar(&c.List, "list", false, "list possible scanners")
+	f.StringVar(&c.Family, "family", "", "family regular expression")
```
Comment (lziest) [inline cli/config.go]: Can we use "scanner family regular expression" for the flag usage string?
